target_chains/sui/contracts/sources/pyth_accumulator.move

```diff
@@ -0,0 +1,116 @@
+module pyth::accumulator {
+    use std::vector::{Self};
+    use sui::clock::{Clock, Self};
+    use wormhole::bytes20::{Self, Bytes20};
+    use wormhole::cursor::{Self, Cursor};
+    use pyth::deserialize::{Self};
+    use pyth::price_identifier::{Self};
+    use pyth::price_info::{Self, PriceInfo};
+    use pyth::price_feed::{Self};
+    use pyth::merkle_tree::{Self};
+
+    const PRICE_FEED_MESSAGE_TYPE: u64 = 0;
+    const E_INVALID_UPDATE_DATA: u64 = 245;
+    const E_INVALID_PROOF: u64 = 345;
+    const E_INVALID_WORMHOLE_MESSAGE: u64 = 454;
+    const E_INVALID_ACCUMULATOR_PAYLOAD: u64 = 554;
+
+    const ACCUMULATOR_UPDATE_WORMHOLE_VERIFICATION_MAGIC: u32 = 1096111958;
+    const PYTHNET_ACCUMULATOR_UPDATE_MAGIC: u64 = 1347305813;
+
+    const MINIMUM_SUPPORTED_MINOR_VERSION: u8 = 0;
+    const MAJOR_VERSION: u8 = 1;
+
+    friend pyth::pyth;
+    #[test_only]
+    friend pyth::pyth_tests;
+
+    // parse_and_verify_accumulator_message verifies that the price updates encoded in the
+    // accumulator message (accessed via cursor) belong to the merkle tree defined by the merkle root encoded in
+    // vaa_payload.
+    public(friend) fun parse_and_verify_accumulator_message(cursor: &mut Cursor<u8>, vaa_payload: vector<u8>, clock: &Clock): vector<PriceInfo> {
+        let major = deserialize::deserialize_u8(cursor);
+        assert!(major == MAJOR_VERSION, E_INVALID_ACCUMULATOR_PAYLOAD);
+
+        let minor = deserialize::deserialize_u8(cursor);
+        assert!(minor >= MINIMUM_SUPPORTED_MINOR_VERSION, E_INVALID_ACCUMULATOR_PAYLOAD);
+
+        let trailing_size = deserialize::deserialize_u8(cursor);
+        deserialize::deserialize_vector(cursor, (trailing_size as u64));
+
+        let proof_type = deserialize::deserialize_u8(cursor);
+        assert!(proof_type == 0, E_INVALID_ACCUMULATOR_PAYLOAD);
+
+        // Ignore the vaa in the accumulator message because presumably it has already been verified
+        // and passed to this function as input.
+        let vaa_size = deserialize::deserialize_u16(cursor);
+        deserialize::deserialize_vector(cursor, (vaa_size as u64));
+
+        let merkle_root_hash = parse_accumulator_merkle_root_from_vaa_payload(vaa_payload);
+        parse_and_verify_accumulator_updates(cursor, merkle_root_hash, clock)
+    }
+
+    // parse_accumulator_merkle_root_from_vaa_payload takes in some VAA bytes, verifies that the vaa
+    // corresponds to a merkle update, and finally returns the keccak hash of the merkle root.
```

Comment:
nitpick: comment here is slightly confusing. "keccak hash of the merkle root" sounds like it does an extra hash of the root for some reason. I think "merkle root" would be clearer.